### ISSUE TITLE
[P3] Accuracy demasiado buena ?

### ISSUE THREAD

**Issue by EleaTriesToCode** [IIC2613-Inteligencia-Artificial-2022-2/Syllabus#65]
Después de limpiar y transformar los datos de acuerdo a las indicaciones del enunciado, he intentado entrenar mi arbol de decicion... 

- Me aparecio que los arboles de decision no pueden manejar valores NaN, lo que me parece raro porque en el ejamplo dado si parece que pueden, y que sin eso la actividad 1 no tiene tanto interes... No sé muy bien que pasa aqui, y que deberia hacer entonces ?

- Para ver que pasa, he suprimido las valores NaN y calculé el accuracy... y obtuve un accuracy de 1.0 !
Me parece muy extrano : no puede ser overfitting porque los datos de test son distintos no ? No entiendo como es posible lograr tal resultado, y que podria ser la fuente de error ? 

Para intentar entender mejor lo que fue pasando, he intentado entrenar un arbol de decision con los datos originales, sin mas modifciaciones que lo necesario para hacer funcionar el arbol (transformacion en numeros y supresion de los NaN) : 

import pandas as pd

data = pd.read_csv("db_diabetes.csv")

data = pd.get_dummies(data, columns=["Diabetes_binary","Sex","HairColor"])

data.dropna(inplace=True)

X = data.loc[:, data.columns != "Diabetes_binary_diabetes"]
y = data["Diabetes_binary_diabetes"]

import sklearn
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)

print(f"Tamaño X train {X_train.shape}")
print(f"Tamaño y train {y_train.shape}")
print(f"Tamaño X test  {X_test.shape}")
print(f"Tamaño y test  {y_test.shape}")

print(f"Diabetes en set entrenamiento: {y_train.sum()/y_train.count()}")
print(f"Diabetes en set test: {y_test.sum()/y_test.count()}")
print(f"Diabetes en todo el set: {y.sum()/y.count()}")

from sklearn.tree import DecisionTreeClassifier
tree = DecisionTreeClassifier(criterion="entropy", max_depth=4)
tree.fit(X_train, y_train)

y_pred = tree.predict(X_test)

from sklearn import metrics

print(metrics.accuracy_score(y_test, y_pred))

==> eso viene por supuesto del ejamplo dado sobre el Titanic, pero me da un resultado también muy extrano : 
Tamaño X train (46509, 31)
Tamaño y train (46509,)
Tamaño X test  (19933, 31)
Tamaño y test  (19933,)
Diabetes en set entrenamiento: 0.5016233417188071
Diabetes en set test: 0.49209853007575377
Diabetes en todo el set: 0.49876584088377834
0.9960868910851352

El accuracy, aun con datos no procesados es super alta ! Me parece entonces que tengo un problema de comprension del método - seria posible ayudarme a identificar lo que es que estoy haciendo mal ?

**Comment by BrunoFarfan** [comment]
Hola!

Viendo el código que pusiste no se ve nada raro. Igualmente recuerda que lo más importante en esta parte de la tarea es el tratamiento de datos (pasarlos a int, eliminar columnas inútiles o **redundantes** y eliminar NaNs).

Si hiciste todo lo que pide el enunciado, no creo que estés haciendo nada mal y es posible que obtengas un valor de accuracy muy alto

**Comment by EleaTriesToCode** [comment]
Gracias por su repuesta !
Creo que si habia algo raro pero si lo puede confirmarle seria super :
No hay duplicados en los datos creo (bueno hay redudancia entre ciertas columnas pero no duplicados exactos) asi que los sets de test y entrenamiento son independiente y el problema viene de otro lugar creo...
Cuando numerizamos las columnas de diabetes_binary, tenemos dos columnas diabetes_binary_diabetes y diabetes_binary_no_diabetes. Me parece que si no eliminamos (como en el ejemplo de arriba) una de las dos columnas existe la posibilidad para el arbol de aprendre algo super facil como "si en diabetes_binary_no_diabetes es 0 en diabetes_binary_diabetes es 1" y asi ser demasiado bueno en los tests. Basicamente, me parece que asi, sin eliminar una de los dos columnas, damos la repuesta al algoritmo y después no extranamos de que sea super bueno :0 
He eliminado una de las dos columnas y entrenado de nuevo el arbol y funciona mejor creo. 

Le parece que si eso fue un problema que tenia ? O fue normal ?

**Comment by BrunoFarfan** [comment]
Creo que entendí que estás usando las columnas que contienen info de si la persona tiene diabetes o no como parte del X_train. Está información son las etiquetas, por lo que no debe estar en la matriz X sino que en el vector y. Si es que pones esta información en X sería como decirle al árbol "tiene diabetes entonces tiene diabetes", con lo que al final no estás haciendo ninguna predicción.

Algo que no entendí bien es ¿creaste 2 columnas a partir de la columna `Diabetes_binary`?

**Comment by EleaTriesToCode** [comment]
Gracias por responder :D
Entiendo lo que me explica sobre el uso de las etiquetas en el train :)
Hace perfecto sentido, es solamente que usando el 
data = pd.get_dummies(data, columns=["Diabetes_binary"])
se crean dos columnas "simetricas" diabetes_binary_diabetes y diabetes_binary_no_diabetes. Asi que depués al usar 
X = data.loc[:, data.columns != "Diabetes_binary_diabetes"]
y = data["Diabetes_binary_diabetes"] 
como en el ejemplo de la ayudantia va a causar el uso de la etiqueta diabetes_binary_no_diabetes en el test y darnos una accuracy super alta. 
Es un error muy tonta pero como eso no fue un problema en el ejmplo de la ayudantia (ya que la columna "Survived" fue siempre unica) me parece que es facil olvidar este detalle y  causarse pb de accurary de 1 :)

**Comment by BrunoFarfan** [comment]
Ahhh ya!! Entonces ya está aclarada la duda. Mucha suerte en el resto de la tarea!
